In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from matplotlib import pyplot as plt


In [2]:
df=pd.read_excel(io='sales_rep.xlsx',sheet_name='sales_data_sample')
df2=pd.read_excel(io='sales_rep.xlsx',sheet_name='Sales_rep',)

In [3]:
df.head(2)

SALES_REP_ID  ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER  \
0             1        10223               37      100.0                1   
1             1        10309               41      100.0                5   

     SALES        ORDERDATE   STATUS  PRODUCTLINE  MSRP PRODUCTCODE  \
0  3965.66   2/20/2004 0:00  Shipped  Motorcycles    95    S10_1678   
1  4394.38  10/15/2004 0:00  Shipped  Motorcycles    95    S10_1678   

                 CUSTOMERNAME       CITY POSTALCODE    COUNTRY DEALSIZE  
0  Australian Collectors, Co.  Melbourne       3004  Australia   Medium  
1          Baane Mini Imports    Stavern       4110     Norway   Medium

In [4]:
df2.head(2)

SALES_REP_ID               Name First Name  Last Name Gender  \
0             1  Harmani Calderoni    Harmani  Calderoni      M   
1             2   Singleton Ashton  Singleton     Ashton      M   

                                           Photo Img  
0  https://images.pexels.com/photos/91227/pexels-...  
1  https://images.pexels.com/photos/718261/pexels...

In [5]:
df=df.iloc[:,1:].merge(df2.iloc[:,:2],left_on='ORDERLINENUMBER',right_on='SALES_REP_ID')

In [6]:
df.head(10)

ORDERNUMBER  QUANTITYORDERED  PRICEEACH  ORDERLINENUMBER    SALES  \
0        10223               37     100.00                1  3965.66   
1        10318               46      94.74                1  4358.04   
2        10210               34     100.00                1  6123.40   
3        10377               50     100.00                1  5182.00   
4        10419               37     100.00                1  5202.94   
5        10405               55     100.00                1  8289.05   
6        10392               36     100.00                1  4035.96   
7        10240               37     100.00                1  5959.22   
8        10105               50      79.67                1  3983.50   
9        10317               35      83.32                1  2916.20   

             ORDERDATE   STATUS   PRODUCTLINE  MSRP PRODUCTCODE  \
0       2/20/2004 0:00  Shipped   Motorcycles    95    S10_1678   
1  2004-02-11 00:00:00  Shipped   Motorcycles    95    S10_1678   
2  2004-12-01 00:00:00  Shipped   Motorcycles   193    S10_4698   
3  2005-09-02 00:00:00  Shipped  Classic Cars   115    S12_4675   
4       5/17/2005 0:00  Shipped  Classic Cars   124    S18_1589   
5       4/14/2005 0:00  Shipped  Classic Cars   169    S18_3232   
6  2005-10-03 00:00:00  Shipped  Vintage Cars    99    S18_3320   
7       4/13/2004 0:00  Shipped  Classic Cars   141    S18_3685   
8  2003-11-02 00:00:00  Shipped  Vintage Cars    83    S24_3816   
9  2004-02-11 00:00:00  Shipped        Planes    72    S24_4278   

                 CUSTOMERNAME        CITY POSTALCODE    COUNTRY DEALSIZE  \
0  Australian Collectors, Co.   Melbourne       3004  Australia   Medium   
1       Diecast Classics Inc.   Allentown      70267        USA   Medium   
2        Osaka Souveniers Co.       Osaka   530-0003      Japan   Medium   
3        Toys of Finland, Co.    Helsinki      21240    Finland   Medium   
4       Salzburg Collectables    Salzburg       5020    Austria   Medium   
5                 Mini Caravy  Strasbourg      67000     France    Large   
6             Mini Auto Werke        Graz       8010    Austria   Medium   
7        Osaka Souveniers Co.       Osaka   530-0003      Japan   Medium   
8    Danish Wholesale Imports   Kobenhavn       1734    Denmark   Medium   
9        Technics Stores Inc.  Burlingame      94217        USA    Small   

   SALES_REP_ID               Name  
0             1  Harmani Calderoni  
1             1  Harmani Calderoni  
2             1  Harmani Calderoni  
3             1  Harmani Calderoni  
4             1  Harmani Calderoni  
5             1  Harmani Calderoni  
6             1  Harmani Calderoni  
7             1  Harmani Calderoni  
8             1  Harmani Calderoni  
9             1  Harmani Calderoni

In [7]:
print(df.shape)
df.info()

(2823, 17)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 2823 entries, 0 to 2822
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ORDERNUMBER      2823 non-null   int64  
 1   QUANTITYORDERED  2823 non-null   int64  
 2   PRICEEACH        2823 non-null   float64
 3   ORDERLINENUMBER  2823 non-null   int64  
 4   SALES            2823 non-null   float64
 5   ORDERDATE        2823 non-null   object 
 6   STATUS           2823 non-null   object 
 7   PRODUCTLINE      2823 non-null   object 
 8   MSRP             2823 non-null   int64  
 9   PRODUCTCODE      2823 non-null   object 
 10  CUSTOMERNAME     2823 non-null   object 
 11  CITY             2823 non-null   object 
 12  POSTALCODE       2747 non-null   object 
 13  COUNTRY          2823 non-null   object 
 14  DEALSIZE         2823 non-null   object 
 15  SALES_REP_ID     2823 non-null   int64  
 16  Name             2823 non-null   object 
dtypes: 

### The Data Structure
The data comprise 2823 rows and 17 columns (features). 6 of these 17 columns (35.3%) are numeric variables

The DEALSIZE is the only ordered factor variable (Small->Medium->Large)

The rest of the variables are non-numeric and they are self-explanatory from their column's name

## Data Accessment
***

## Data Cleaning
***

In [8]:
x=[x.lower() for x in df.columns ]

In [9]:
# paste and edit output x
x=['order_number',
 'quantity_ordered',
 'price_each',
 'order_line_number',
 'sales',
 'order_date',
 'status',
 'product_line',
 'msrp',
 'product_code',
 'customer_name',
 'city',
 'postal_code',
 'country',
 'deal_size',
 'sales_rep_id',
 'name']
# replace data frame's columns with x
df.columns=x

In [10]:
df.head(2)

order_number  quantity_ordered  price_each  order_line_number    sales  \
0         10223                37      100.00                  1  3965.66   
1         10318                46       94.74                  1  4358.04   

            order_date   status product_line  msrp product_code  \
0       2/20/2004 0:00  Shipped  Motorcycles    95     S10_1678   
1  2004-02-11 00:00:00  Shipped  Motorcycles    95     S10_1678   

                customer_name       city postal_code    country deal_size  \
0  Australian Collectors, Co.  Melbourne        3004  Australia    Medium   
1       Diecast Classics Inc.  Allentown       70267        USA    Medium   

   sales_rep_id               name  
0             1  Harmani Calderoni  
1             1  Harmani Calderoni

In [11]:
cost_of_goods=df.quantity_ordered*df.price_each
df.insert(3,column='cost_of_goods',value=cost_of_goods)

In [12]:
profit=df.sales-round(df.cost_of_goods,2)
df.insert(6,column='profit',value=profit)

In [13]:
print('The data frame conatins {} duplicate(s)'.format (df.duplicated().sum()))

The data frame conatins 0 duplicate(s)


## Exploratory Data Analysis
***

In [14]:
df.describe()

order_number  quantity_ordered   price_each  cost_of_goods  \
count   2823.000000       2823.000000  2823.000000    2823.000000   
mean   10258.725115         35.092809    83.658544    2936.906408   
std       92.085478          9.741443    20.174277    1105.644284   
min    10100.000000          6.000000    26.880000     482.130000   
25%    10180.000000         27.000000    68.860000    2100.000000   
50%    10262.000000         35.000000    95.700000    2800.000000   
75%    10333.500000         43.000000   100.000000    3726.810000   
max    10425.000000         97.000000   100.000000    9048.160000   

       order_line_number         sales       profit         msrp  sales_rep_id  
count        2823.000000   2823.000000  2823.000000  2823.000000   2823.000000  
mean            6.466171   3553.889072   616.982664   100.715551      6.466171  
std             4.225841   1841.865106  1047.728887    40.187912      4.225841  
min             1.000000    482.130000     0.000000    33.000000      1.000000  
25%             3.000000   2203.430000     0.000000    68.000000      3.000000  
50%             6.000000   3184.800000     0.000000    99.000000      6.000000  
75%             9.000000   4508.000000   925.180000   124.000000      9.000000  
max            18.000000  14082.800000  7536.500000   214.000000     18.000000

In [15]:
df

order_number  quantity_ordered  price_each  cost_of_goods  \
0            10223                37      100.00        3700.00   
1            10318                46       94.74        4358.04   
2            10210                34      100.00        3400.00   
3            10377                50      100.00        5000.00   
4            10419                37      100.00        3700.00   
...            ...               ...         ...            ...   
2818         10168                27       73.02        1971.54   
2819         10159                42       51.48        2162.16   
2820         10222                36       63.34        2280.24   
2821         10165                31       71.10        2204.10   
2822         10316                48       74.45        3573.60   

      order_line_number    sales   profit           order_date   status  \
0                     1  3965.66   265.66       2/20/2004 0:00  Shipped   
1                     1  4358.04     0.00  2004-02-11 00:00:00  Shipped   
2                     1  6123.40  2723.40  2004-12-01 00:00:00  Shipped   
3                     1  5182.00   182.00  2005-09-02 00:00:00  Shipped   
4                     1  5202.94  1502.94       5/17/2005 0:00  Shipped   
...                 ...      ...      ...                  ...      ...   
2818                 18  1971.54     0.00      10/28/2003 0:00  Shipped   
2819                 18  2162.16     0.00  2003-10-10 00:00:00  Shipped   
2820                 18  2280.24     0.00       2/19/2004 0:00  Shipped   
2821                 18  2204.10     0.00      10/22/2003 0:00  Shipped   
2822                 18  3573.60     0.00  2004-01-11 00:00:00  Shipped   

          product_line  msrp product_code                 customer_name  \
0          Motorcycles    95     S10_1678    Australian Collectors, Co.   
1          Motorcycles    95     S10_1678         Diecast Classics Inc.   
2          Motorcycles   193     S10_4698          Osaka Souveniers Co.   
3         Classic Cars   115     S12_4675          Toys of Finland, Co.   
4         Classic Cars   124     S18_1589         Salzburg Collectables   
...                ...   ...          ...                           ...   
2818            Planes    68     S24_3949          Technics Stores Inc.   
2819       Motorcycles    60     S18_2625      Corporate Gift Ideas Co.   
2820             Ships    54     S72_3212  Collectable Mini Designs Co.   
2821  Trucks and Buses    60     S18_2432       Dragon Souveniers, Ltd.   
2822             Ships    72    S700_2610             giftsbymail.co.uk   

               city postal_code    country deal_size  sales_rep_id  \
0         Melbourne        3004  Australia    Medium             1   
1         Allentown       70267        USA    Medium             1   
2             Osaka    530-0003      Japan    Medium             1   
3          Helsinki       21240    Finland    Medium             1   
4          Salzburg        5020    Austria    Medium             1   
...             ...         ...        ...       ...           ...   
2818     Burlingame       94217        USA     Small            18   
2819  San Francisco         NaN        USA     Small            18   
2820      San Diego       91217        USA     Small            18   
2821      Singapore       79903  Singapore     Small            18   
2822          Cowes    PO31 7PJ         UK    Medium            18   

                   name  
0     Harmani Calderoni  
1     Harmani Calderoni  
2     Harmani Calderoni  
3     Harmani Calderoni  
4     Harmani Calderoni  
...                 ...  
2818   Theodore Drayers  
2819   Theodore Drayers  
2820   Theodore Drayers  
2821   Theodore Drayers  
2822   Theodore Drayers  

[2823 rows x 19 columns]

In [19]:
df.query('profit >0')

order_number  quantity_ordered  price_each  cost_of_goods  \
0            10223                37       100.0         3700.0   
2            10210                34       100.0         3400.0   
3            10377                50       100.0         5000.0   
4            10419                37       100.0         3700.0   
5            10405                55       100.0         5500.0   
...            ...               ...         ...            ...   
2806         10332                26       100.0         2600.0   
2808         10210                31       100.0         3100.0   
2809         10312                48       100.0         4800.0   
2810         10316                33       100.0         3300.0   
2812         10280                35       100.0         3500.0   

      order_line_number    sales   profit           order_date   status  \
0                     1  3965.66   265.66       2/20/2004 0:00  Shipped   
2                     1  6123.40  2723.40  2004-12-01 00:00:00  Shipped   
3                     1  5182.00   182.00  2005-09-02 00:00:00  Shipped   
4                     1  5202.94  1502.94       5/17/2005 0:00  Shipped   
5                     1  8289.05  2789.05       4/14/2005 0:00  Shipped   
...                 ...      ...      ...                  ...      ...   
2806                 17  2979.08   379.08      11/17/2004 0:00  Shipped   
2808                 17  5719.50  2619.50  2004-12-01 00:00:00  Shipped   
2809                 17  8078.40  3278.40      10/21/2004 0:00  Shipped   
2810                 17  4128.96   828.96  2004-01-11 00:00:00  Shipped   
2812                 17  3704.05   204.05       8/17/2004 0:00  Shipped   

      product_line  msrp product_code                 customer_name  \
0      Motorcycles    95     S10_1678    Australian Collectors, Co.   
2      Motorcycles   193     S10_4698          Osaka Souveniers Co.   
3     Classic Cars   115     S12_4675          Toys of Finland, Co.   
4     Classic Cars   124     S18_1589         Salzburg Collectables   
5     Classic Cars   169     S18_3232                   Mini Caravy   
...            ...   ...          ...                           ...   
2806  Vintage Cars    62     S18_2957                AV Stores, Co.   
2808        Planes   157     S18_1662          Osaka Souveniers Co.   
2809  Vintage Cars   170     S18_1749  Mini Gifts Distributors Ltd.   
2810  Classic Cars   136     S10_4757             giftsbymail.co.uk   
2812  Vintage Cars    92     S18_4409            Amica Models & Co.   

            city postal_code    country deal_size  sales_rep_id  \
0      Melbourne        3004  Australia    Medium             1   
2          Osaka    530-0003      Japan    Medium             1   
3       Helsinki       21240    Finland    Medium             1   
4       Salzburg        5020    Austria    Medium             1   
5     Strasbourg       67000     France     Large             1   
...          ...         ...        ...       ...           ...   
2806  Manchester     EC2 5NT         UK     Small            17   
2808       Osaka    530-0003      Japan    Medium            17   
2809  San Rafael       97562        USA     Large            17   
2810       Cowes    PO31 7PJ         UK    Medium            17   
2812      Torino       10100      Italy    Medium            17   

                   name  
0     Harmani Calderoni  
2     Harmani Calderoni  
3     Harmani Calderoni  
4     Harmani Calderoni  
5     Harmani Calderoni  
...                 ...  
2806        Taylor Jole  
2808        Taylor Jole  
2809        Taylor Jole  
2810        Taylor Jole  
2812        Taylor Jole  

[1304 rows x 19 columns]

0        265.66
1          0.00
2       2723.40
3        182.00
4       1502.94
         ...   
2818       0.00
2819       0.00
2820       0.00
2821       0.00
2822       0.00
Name: profit, Length: 2823, dtype: float64